### ISSUE TITLE
Overriding getSeqlengths

### ISSUE THREAD

**Issue by stevenjblair** [Bioconductor/BSgenome#67]
Hello, I had issues before and you were very helpful!  Now, I have a new problem. 

I work with the axolotl genome, it is huge.  We have performed ATAC seq, in order to do this we needed to cut the chromosomes down to smaller chunks ([as described here](https://github.com/brianjohnhaas/ChunkGenome)).   Picard has a limitation of 2^29 base length.  Anyhow, we did that and it worked great.  I would love to create a new BSgenome, in my seed I simply set seqnames: scaf where scaf is a character vector of a list of the directory of chromosomes, that seems to work fine.  But the issue now is I get this error:

> Error in getSeqlengths(seqnames, prefix = prefix, suffix = suffix, seqs_srcdir = seqs_srcdir,  : 
>   the sequences in the files have lengths that don't match the lengths reported by 'Seqinfo(genome="AmbMex60DD")'

Can we just detach this from NCBI all together?  It seems to be erroring only because it is different than the original genome on the server.

Best regards

**Comment by stevenjblair** [comment]
My seed

> Package: BSgenome.Amexicanum.NCBI.ambMex60DD.chunked
> Title: Chunked genome sequences for Ambystoma mexicanum (AmbMex60DD)
> Description: Full genome sequences for Ambystoma mexicanum (Axolotl) as provided by NCBI (assembly AmbMex60DD, assembly accession GCA_002915635.3) and stored in Biostrings objects.
> Version: 1.0.0c
> organism: Ambystoma mexicanum
> common_name: Axolotl
> genome: AmbMex60DD
> provider: NCBI
> release_date: 2021/04/01
> source_url: https://www.ncbi.nlm.nih.gov/assembly/GCA_002915635.3
> organism_biocview: Ambystoma_mexicanum
> BSgenomeObjname: Amexicanum
> seqnames: scaf
> circ_seqs: character(0)
> SrcDataFiles: GCA_002915635.3_AmbMex60DD_genomic.fna.gz from https://ftp.ncbi.nlm.nih.gov/genomes/all/GCA/002/915/635/GCA_002915635.3_AmbMex60DD/
> PkgExamples: genome$chr1p^c0^o1  # same as genome[["chr1p^c0^o1"]]
> seqs_srcdir: /n/holyscratch01/whited_lab/sjblair/Projects/BSgenome_chunkAmex60DD/scaf
> seqfiles_suffix: .fna
> ondisk_seq_format: rds

**Comment by hpages** [comment]
Use another name for the genome. If you call this `AmbMex60DD` then some tools in Bioconductor will be confused and break because they'll expect to see the original AmbMex60DD sequences in this BSgenome data package, but that's not what it contains.

So you could just make up your own name for this genome e.g. `AmbMex60DDchunked`. Note that you should also use that in the package name (`BSgenome.Amexicanum.NCBI.AmbMex60DDchunked`). Also since you are now forging a BSgenome data package for a genome that is not registered, then you'll need to list all the sequence names in the `seqnames` field (make sure to provide them as an R character vector e.g. `seqnames: c("seq1", "seq2", etc...)`). You'll also need to provide one FASTA file per pseudo-sequence (i.e. for each sequence chunk).

I see that you're using `genome$chr1p^c0^o1` in the `PkgExamples` field. Are you sure you want to use this kind of names for your pseudo-sequences? This sounds that it will likely give your users a hard time!

**Comment by hpages** [comment]
Were you able to sort this out?

**Comment by stevenjblair** [comment]
Hi Hervé, I vanished for awhile.  We decided to go another way but now I am in the same place trying to finish this.  I did as you said, I changed the name, and then I added all ~27k names in a character vector.  I have not changed the scaffold names yet as we have a lot of files that we will need to use that have that naming convention.  When I run `forgeBSgenomeDataPkg("./ambmex60DDchunked.seed", seqs_srcdir = "./fna", destdir = "./BSgenome")` I get the following error
`Error in h(simpleError(msg, call)) : 
  error in evaluating the argument 'x' in selecting a method for function 'seqlengths': "ambMex60DD.chunked" is not a registered NCBI assembly or UCSC genome (use registered_NCBI_assemblies() or registered_UCSC_genomes() to
  list the NCBI or UCSC assemblies/genomes currently registered in the GenomeInfoDb package)
In addition: Warning message:
In readLines(infile, n = 25000L) :
  incomplete final line found on './ambmex60DDchunked.seed'`

I have the seqlengths (I am assuming they are the lengths of the fasta files), I'm just unsure how to point to them.  

Any ideas?

**Comment by stevenjblair** [comment]
Here is my seed file, I removed a bunch of the names to make it more readable
`## seed file start here
Package: BSgenome.Amexicanum.NCBI.ambMex60DD.chunked
Title: Chunked genome sequences for Ambystoma mexicanum (AmbMex60DD)
Description: Modified genome sequences for Ambystoma mexicanum (Axolotl), chunked into smaller sequences and stored in Biostrings objects.
Version: 1.0.0c
organism: Ambystoma mexicanum
common_name: Axolotl
genome: ambMex60DD.chunked
provider: NCBI
release_date: 2021/04/01
source_url: ./fna/
organism_biocview: Ambystoma_mexicanum
BSgenomeObjname: Amexicanum
seqnames: c("C0000568",...,"chr10p^c1^o360766498","chr10p^c2^o721257583","chr10q^c0^o1","chr10q^c1^o263007889","chr11p","chr11q^c0^o1","chr11q^c1^o358409797","chr11q^c2^o719828585","chr12p","chr12q^c0^o1","chr12q^c1^o428970560","chr13p","chr13q","chr14p","chr14q","chr1p^c0^o1","chr1p^c1^o472790116","chr1p^c2^o945337135","chr1q^c0^o1","chr1q^c1^o490778492","chr1q^c2^o981659532","chr2p^c0^o1","chr2p^c1^o460811079","chr2p^c2^o921610769","chr2q^c0^o1","chr2q^c1^o471547657","chr2q^c2^o943406818","chr3p^c0^o1","chr3p^c1^o433958941","chr3q^c0^o1","chr3q^c1^o381863112","chr3q^c2^o763400146","chr3q^c3^o1146032874","chr4p^c0^o1","chr4p^c1^o402265070","chr4p^c2^o803787347","chr4q^c0^o1","chr4q^c1^o404240860","chr4q^c2^o810769300","chr5p^c0^o1","chr5p^c1^o426853249","chr5p^c2^o853659530","chr5q^c0^o1","chr5q^c1^o428721598","chr5q^c2^o856566866","chr6p^c0^o1","chr6p^c1^o487623146","chr6p^c2^o975176869","chr6q^c0^o1","chr6q^c1^o406785210","chr6q^c2^o813579487","chr6q^c3^o1220072138","chr7p^c0^o1","chr7p^c1^o454432061","chr7q^c0^o1","chr7q^c1^o361312010","chr7q^c2^o722752951","chr8p^c0^o1","chr8p^c1^o373558763","chr8q^c0^o1","chr8q^c1^o442126268","chr9p","chr9q^c0^o1","chr9q^c1^o334212844","chr9q^c2^o668444789","chrM")
circ_seqs: character(0)
PkgExamples: genome$chr1p^c0^o1 # same as genome[["chr1p^c0^o1"]]
seqs_srcdir: /n/holylabs/LABS/whited_lab/Users/sjblair/ATAC_chromVAR_Mar212024/fna
seqfiles_suffix: .fna
ondisk_seq_format: rds
## seed file end here ---`

**Comment by hpages** [comment]
Hi Steven,

So you confirm that you have ~27k FASTA files in `/n/holylabs/LABS/whited_lab/Users/sjblair/ATAC_chromVAR_Mar212024/fna/` and that they are named according to your ~27k sequence names? Then this should work.

What's your `sessionInfo()`? Make sure to use the latest version of Bioconductor (3.18) and that all your packages are up-to-date (check this with `BiocManager::valid()`).

Note that everything would be much easier if you were storing all the sequences in a 2bit file. Then you wouldn't have to list the ~27k sequence names in your seed file. Also the resulting BSgenome data package will be smaller and loading the sequences will be much faster.

H.

**Comment by stevenjblair** [comment]
I have checked off everything on your list except the bioconductor and R version.  Our cluster only officially supports up to 3.16.  I will need to install the correct versions on my own.  I will keep you informed on my progress. 

Thanks again so much

**Comment by hpages** [comment]
Bioconductor 3.16 is an old and unsupported version. People should always use the current version (currently 3.18 but 3.19 will be released in about a month). Someone really needs to talk to your cluster admins.

**Comment by stevenjblair** [comment]
Don't get me started on that, haha.  I have installed 3.18 and R-4.3.  I might have a go at this tonight or tomorrow.  

I did split the whole genome fasta into single fasta contig files last night and attempted to use the 2bit method.  It seems to fail once the 2bit file gets to 4 Gb.  I thought it would turn each fasta into it's own 2bit, but that was not the behavior I observed.  Perhaps it would work with 3.18 though.  We will see.
